docs/design_notes/lift.md

```diff
@@ -0,0 +1,340 @@
+⚠️ Advanced topic ⚠️
+
+This design note explains the underlying implementation of `flax.linen.transform`, which enables JAX transformations inside `Module`.
+
+
+## Introduction
+
+JAX uses a functional API meaning that it only guarantees correct behavior when using functions without side effects.
+Typically, these side effects are the result of mutating an object that lives outside the function.
+
+The functional paradigm has some advantages like the ability to explicitly reason about state and stochasticity.
+The function output only changes when an input argument changes.
+Therefore, a function is guaranteed to behave deterministically.
+
+But pure functions offer another big advantage to JAX: specifically, they enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
+Would `f` draw the same or a different random number for each item in the vector?
+Would each item in the batch have its own counter or is the counter shared among the items?
+And in what order is the counter incremented if `f` is computed in parallel?
+The answer to all these questions is "it depends".
+The behavior is ambiguous and the functional constraint elegantly avoids this problem.
+
+Flax introduces a safe way to have limited randomness and stateful variables in a JAX compatible form.
+The reason why the state in Flax is not problematic is because it is local: inside a Flax `Module` there are variables and PRNG sequences,
+but on the outside there are only JAX Arrays and PRNG keys.
+
+For most use cases, Flax is used to define models in a stateful way.
+Because a `Module` behaves like a pure function externally, we can fully utilize JAX with all of its transformations.
+There are, however, cases when we want to have the best of both worlds by using transformations and `Module` together.
+This design note explains how we extend JAX's functional transformation to work on `Module`s that have internal state and randomness.
+
+
+## Functionalization
+
+Before we jump into the details let's consider a simple example where we would like to use `vmap` inside a `Module`.
+
+First, we define a simple MLP without any transformations:
+
+```python
+import jax
+from jax import random, NumPy as jnp
+from flax import linen as nn
+
+class MLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    h = nn.Dense(4, name='hidden')(xs)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+```
+
+Now what if we want to have separate MLP parameters for each item in `xs`?
+If this were "vanilla JAX" we could imagine writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
+But doing something like this in Linen will actually fail:
+
+```python
+class NaiveVmapMLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    mlps = MLP()
```

Comment:
maybe `mlps = [MLP() for _ in xs.shape[0]]` ?
(depends what you want to highlight - but it looks weird having plural in the var name with a singular class name)

Reply:
I was thinking mlps because you vmap the MLP which is equivalent to applying a bunch of mlps in parallel. But maybe that's a bit confusing. I'll remove the plural here